### PR TITLE
Support for multi agent mode of training and custom traffic agents

### PR DESCRIPTION
**Pull request by buridiaditya**
A flag is used to identify to the mode of training (single or multi agent). This is necessary as in single agent training every instance of the Environment creates a new simulator instance which is not desired in multi agent. When the flag `multi_agent_mode` is set to True simulator instance is not created.

For custom traffic a list of agent numbers is passed to Env. This list determines the traffic agent type. Custom traffic is needed for training agent 

Testing:
For multi agent:
Ran torcs with 4 scr_server players in quick race and connected 4 agents by creating 4 environment instance and each calling reset().

For traffic:
Just run a single agent and set the traffic list to [1,1,1]. Make sure to set enough scr_server agents in torcs simulator.

### CODE REVIEW

**Comment by rudrasohan** [comment]
@buridiaditya for multi-agent mode do we need to run two algos separately or an instance of mpi works ? I tried using the mpi algo ddpg of baselines but kept on getting timeouts.

**Review by rudrasohan — COMMENTED** [review]
While importing traffic agent's there is a slight problem with the imports of various sub packages of madras.
My baselines folder is not in the same area as my Madras folder.

File: MADRaS/traffic/const_vel.py [hunk 1]
```diff
@@ -2,17 +2,17 @@
...
-from utils.gym_torcs import TorcsEnv
-import utils.snakeoil3_gym as snakeoil3
-from utils.madras_datatypes import Madras
+from MADRaS.controllers.pid import PID
+from MADRaS.utils.gym_torcs import TorcsEnv
+import MADRaS.utils.snakeoil3_gym as snakeoil3
+from MADRaS.utils.madras_datatypes import Madras
 
 madras = Madras()
 with open("./traffic/configurations.yml", "r") as ymlfile:
```
Comment (rudrasohan) [inline MADRaS/traffic/const_vel.py]: There is a small issue is with this open statement. I needed to specify my complete path from home for each of these for the file to be found.

Comment (buridiaditya) [reply MADRaS/traffic/const_vel.py]: yeah I kept testing from the MADRaS folder.

File: MADRaS/traffic/const_vel.py [hunk 2]
```diff
@@ -2,17 +2,17 @@
 import sys
```
Comment (rudrasohan) [inline MADRaS/traffic/const_vel.py:2]: Also the code was not able to find the imports of lines 
```python
from MADRaS.controllers.pid import PID
from MADRaS.utils.gym_torcs import TorcsEnv
import MADRaS.utils.snakeoil3_gym as snakeoil3
from MADRaS.utils.madras_datatypes import Madras
```

so I had to add the line `sys.path.append("/home/sohan/projects/MADRaS/MADRaS")` for it to work.

Comment (buridiaditya) [reply MADRaS/traffic/const_vel.py:2]: If you install MADRaS using `pip install -e .` from the MADRaS home folder it should work.

**Comment by buridiaditya** [comment]
> @buridiaditya for multi-agent mode do we need to run two algos separately or an instance of mpi works ? I tried using the mpi algo ddpg of baselines but kept on getting timeouts.

For the multiagent mode torcs simulator is not started by the MadrasEnv anymore. This should be manually started by the training algo.

File: MADRaS/traffic/const_vel.py (hunk 2 above)
Comment (rudrasohan) [reply MADRaS/traffic/const_vel.py:2]: I did it using the same way as you suggested. 
Upon removing the line this is the error traceback that I get.
```shell
  File "/home/sohan/anaconda2/envs/madras/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2338, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/sohan/projects/MADRaS/MADRaS/envs/__init__.py", line 2, in <module>
    from MADRaS.envs.gym_madras import MadrasEnv
  File "/home/sohan/projects/MADRaS/MADRaS/envs/gym_madras.py", line 22, in <module>
    import MADRaS.traffic.vel_change as agentSinusoid
  File "/home/sohan/projects/MADRaS/MADRaS/traffic/vel_change.py", line 6, in <module>
    from controllers.pid import PID
ImportError: No module named 'controllers'

```

Comment (rudrasohan) [reply MADRaS/traffic/const_vel.py:2]: Oh sorry,  got the problem

Comment (rudrasohan) [reply MADRaS/traffic/const_vel.py:2]: Actually, the some traffic files do not have their imports correctly set up. [vel_change.py](https://github.com/madras-simulator/MADRaS/blob/63c8c88e6917e26a8e810b1ab68a6014a995b6b4/MADRaS/traffic/vel_change.py#L6-L9)

**Comment by rudrasohan** [comment]
> > @buridiaditya for multi-agent mode do we need to run two algos separately or an instance of mpi works ? I tried using the mpi algo ddpg of baselines but kept on getting timeouts.
> 
> For the multiagent mode torcs simulator is not started by the MadrasEnv anymore. This should be manually started by the training algo.

Could you give a sample on how to start the training.

File: MADRaS/envs/gym_madras.py [hunk 3]
```diff
@@ -123,7 +154,30 @@ def reset(self, prev_step_info=None):
 
         else:
             try:
```
Comment (Santara) [inline MADRaS/envs/gym_madras.py:168]: @buridiaditya add a line here making a random choice of traffic agents if randomize_traffic==True, before the ```try``` block:
```
if self.randomize_traffic:
    self.traffic_type = [np.random.randint(0, 4) for _ in self.traffic_type]
```

File: MADRaS/envs/gym_madras.py [hunk 4]
```diff
@@ -27,15 +32,19 @@
...
 import socket
 
+from multiprocessing import Process
+
 class MadrasEnv(TorcsEnv,gym.Env):
     """Definition of the Gym Madras Env."""
     def __init__(self, vision=False, throttle=True,
                  gear_change=False, port=60934, pid_assist=False,
-                 CLIENT_MAX_STEPS=np.inf,visualise=True,no_of_visualisations=1):
+                 CLIENT_MAX_STEPS=np.inf,visualise=True,no_of_visualisations=1, multi_agent_mode=False , traffic_type=[0,0,0]):
```
Comment (Santara) [inline MADRaS/envs/gym_madras.py]: Add an argument, ```randomize_traffic```

File: MADRaS/envs/gym_madras.py [hunk 5]
```diff
@@ -27,15 +32,19 @@
...
+        # traffic_type is a list of traffic agents.
         # If `visualise` is set to False torcs simulator will run in headless mode
         """Init Method."""
         self.torcs_proc = None
         self.pid_assist = pid_assist
+        self.traffic_type = traffic_type
```
Comment (Santara) [inline MADRaS/envs/gym_madras.py:48]: add after this:
```
self.randomize_traffic = randomize_traffic
```

**Review by Santara — DISMISSED** [review]

**Comment by MehaKaushik** [comment]
> > > @buridiaditya for multi-agent mode do we need to run two algos separately or an instance of mpi works ? I tried using the mpi algo ddpg of baselines but kept on getting timeouts.
> > 
> > 
> > For the multiagent mode torcs simulator is not started by the MadrasEnv anymore. This should be manually started by the training algo.
> 
> Could you give a sample on how to start the training.

Following !!

**Comment by rudrasohan** [comment]
Timeout arising. (race is not starting)

Steps to reproduce:
- Set up torcs with 4 scr agents
- Test code
```python
import MADRaS
import gym


env = gym.make("Madras-v0")
ob = env.reset()
r_t = 0
for i in range(1500):
    ob, r, done, _ = env.step([0, 0.2])
    r_t = r_t+r
    print("STEP={},REW={},TOTAL={}".format(i,r,r_t))
    if done:
    	print("DONE DONE")
    	break
```
- set traffic agents to `[1,1,1]`
image

**Comment by rudrasohan** [comment]
Merging and work continued on #30

**Pull request merged**